File: backend/src/integration/bmc/redfish.rs
//! Redfish BMC client (HTTPS + basic auth).

use reqwest::header::CONTENT_TYPE;
use serde_json::Value;
use std::time::Duration;

use super::{BootTarget, BmcCredentials, PowerState};
use crate::AppError;

pub struct RedfishClient {
    http: reqwest::Client,
    base: String,
    system_path: String,
    username: String,
    password: String,
}

impl RedfishClient {
    pub async fn connect(creds: &BmcCredentials) -> Result<Self, AppError> {
        let host = creds
            .address
            .trim_start_matches("https://")
            .trim_start_matches("http://")
            .trim_end_matches('/');
        let base = format!("https://{host}");

        let http = reqwest::Client::builder()
            .timeout(Duration::from_secs(creds.timeout_secs.max(5)))
            .danger_accept_invalid_certs(creds.tls_insecure)
            .user_agent("tcs-bmc/0.3")
            .build()
            .map_err(|e| AppError::Network(format!("Redfish client build: {e}")))?;

        let systems_url = format!("{base}/redfish/v1/Systems");
        let resp = http
            .get(&systems_url)
            .basic_auth(&creds.username, Some(&creds.password))
            .send()
            .await
            .map_err(|e| AppError::Network(format!("Redfish Systems: {e}")))?;

        if !resp.status().is_success() {
            return Err(AppError::Network(format!(
                "Redfish Systems HTTP {}",
                resp.status()
            )));
        }

        let body: Value = resp
            .json()
            .await
            .map_err(|e| AppError::Network(format!("Redfish Systems JSON: {e}")))?;

        let system_path = if !creds.redfish_path.trim().is_empty() {
            let p = creds.redfish_path.trim();
            if p.starts_with('/') {
                p.to_string()
            } else {
                format!("/{p}")
            }
        } else {
            body.get("Members")
                .and_then(|m| m.as_array())
                .and_then(|arr| arr.first())
                .and_then(|m| m.get("@odata.id"))
                .and_then(|v| v.as_str())
                .map(|s| s.to_string())
                .ok_or_else(|| AppError::Network("Redfish: no Systems members found".into()))?
        };

        Ok(Self {
            http,
            base,
            system_path,
            username: creds.username.clone(),
            password: creds.password.clone(),
        })
    }

    fn system_url(&self) -> String {
        if self.system_path.starts_with("http") {
            self.system_path.clone()
        } else {
            format!("{}{}", self.base, self.system_path)
        }
    }

    pub async fn get_power_state(&self) -> Result<PowerState, AppError> {
        let resp = self
            .http
            .get(self.system_url())
            .basic_auth(&self.username, Some(&self.password))
            .send()
            .await
            .map_err(|e| AppError::Network(format!("Redfish get system: {e}")))?;
        if !resp.status().is_success() {
            return Err(AppError::Network(format!(
                "Redfish get system HTTP {}",
                resp.status()
            )));
        }
        let body: Value = resp
            .json()
            .await
            .map_err(|e| AppError::Network(format!("Redfish system JSON: {e}")))?;
        let state = body
            .get("PowerState")
            .and_then(|v| v.as_str())
            .unwrap_or("Unknown");
        Ok(match state {
            "On" => PowerState::On,
            "Off" => PowerState::Off,
            _ => PowerState::Unknown,
        })
    }

    pub async fn power(&self, action: &str) -> Result<(), AppError> {
        let reset_type = match action {
            "on" => "On",
            "off" => "ForceOff",
            "reset" | "cycle" => "ForceRestart",
            "graceful_shutdown" => "GracefulShutdown",
            other => {
                return Err(AppError::InvalidInput(format!(
                    "Unknown power action: {other}"
                )))
            }
        };
        let url = format!("{}/Actions/ComputerSystem.Reset", self.system_url());
        let body = serde_json::json!({ "ResetType": reset_type });
        let resp = self
            .http
            .post(&url)
            .basic_auth(&self.username, Some(&self.password))
            .json(&body)
            .send()
            .await
            .map_err(|e| AppError::Network(format!("Redfish power: {e}")))?;
        let code = resp.status().as_u16();
        if resp.status().is_success() || code == 204 || code == 202 {
            return Ok(());
        }
        let text = resp.text().await.unwrap_or_default();
        Err(AppError::Network(format!(
            "Redfish power HTTP {code}: {text}"
        )))
    }

    pub async fn mount_iso(&self, iso_url: &str, media: &str) -> Result<(), AppError> {
        let vmedia_path = format!("{}/VirtualMedia", self.system_url());
        let resp = self
            .http
            .get(&vmedia_path)
            .basic_auth(&self.username, Some(&self.password))
            .send()
            .await
            .map_err(|e| AppError::Network(format!("Redfish get VirtualMedia: {e}")))?;
        if !resp.status().is_success() {
            return Err(AppError::Network(format!(
                "Redfish get VirtualMedia HTTP {}",
                resp.status()
            )));
        }
        let body: Value = resp
            .json()
            .await
            .map_err(|e| AppError::Network(format!("Redfish VirtualMedia JSON: {e}")))?;
        let members = body.get("Members")
            .and_then(|m| m.as_array())
            .ok_or_else(|| AppError::Network("No VirtualMedia members found".into()))?;
        let mut target_member: Option<&Value> = None;
        let mut target_odata: Option<String> = None;
        for m in members {
            let name = m.get("Name")
                .and_then(|n| n.as_str())
                .unwrap_or("");
            let odata = m.get("@odata.id")
                .and_then(|n| n.as_str())
                .unwrap_or("");
            if name.to_lowercase() == media.to_lowercase() {
                target_member = Some(m);
                target_odata = Some(odata.to_string());
                break;
            }
        }
        if target_member.is_none() {
            return Err(AppError::Network(format!(
                "VirtualMedia '{media}' not found. Available: {:?}",
                members.iter().filter_map(|m| m.get("Name").and_then(|n| n.as_str())).collect::<Vec<_>>()
            )));
        }
        let detail_path = target_odata.unwrap();
        let insert_url = format!("{}/VirtualMedia.InsertMedia",
            if detail_path.starts_with("http") {
                detail_path
            } else {
                format!("{}{}", self.base, detail_path)
            }
        );
        let body = serde_json::json!({
            "Image": iso_url,
            "TransferMethod": "Download",
            "TransferProtocolType": "HTTP"
        });
        let resp = self
            .http
            .post(&insert_url)
            .basic_auth(&self.username, Some(&self.password))
            .header(CONTENT_TYPE, "application/json")
            .json(&body)
            .send()
            .await
            .map_err(|e| AppError::Network(format!("Redfish mount ISO: {e}")))?;
        let code = resp.status().as_u16();
        if resp.status().is_success() || code == 204 || code == 202 {
            return Ok(());
        }
        let text = resp.text().await.unwrap_or_default();
        Err(AppError::Network(format!(
            "Redfish mount ISO HTTP {code}: {text}"
        )))
    }

    pub async fn unmount_iso(&self, media: &str) -> Result<(), AppError> {
        let vmedia_path = format!("{}/VirtualMedia", self.system_url());
        let resp = self
            .http
            .get(&vmedia_path)
            .basic_auth(&self.username, Some(&self.password))
            .send()
            .await
            .map_err(|e| AppError::Network(format!("Redfish get VirtualMedia: {e}")))?;
        if !resp.status().is_success() {
            return Err(AppError::Network(format!(
                "Redfish get VirtualMedia HTTP {}",
                resp.status()
            )));
        }
        let body: Value = resp
            .json()
            .await
            .map_err(|e| AppError::Network(format!("Redfish VirtualMedia JSON: {e}")))?;
        let members = body.get("Members")
            .and_then(|m| m.as_array())
            .ok_or_else(|| AppError::Network("No VirtualMedia members found".into()))?;
        let mut target_odata: Option<String> = None;
        for m in members {
            let name = m.get("Name")
                .and_then(|n| n.as_str())
                .unwrap_or("");
            let odata = m.get("@odata.id")
                .and_then(|n| n.as_str())
                .unwrap_or("");
            if name.to_lowercase() == media.to_lowercase() {
                target_odata = Some(odata.to_string());
                break;
            }
        }
        let detail_path = target_odata.ok_or_else(|| AppError::Network(format!(
            "VirtualMedia '{media}' not found"
        )))?;
        let eject_url = format!("{}/VirtualMedia.EjectMedia",
            if detail_path.starts_with("http") {
                detail_path
            } else {
                format!("{}{}", self.base, detail_path)
            }
        );
        let resp = self
            .http
            .post(&eject_url)
            .basic_auth(&self.username, Some(&self.password))
            .header(CONTENT_TYPE, "application/json")
            .send()
            .await
            .map_err(|e| AppError::Network(format!("Redfish unmount ISO: {e}")))?;
        let code = resp.status().as_u16();
        if resp.status().is_success() || code == 204 || code == 202 {
            return Ok(());
        }
        let text = resp.text().await.unwrap_or_default();
        Err(AppError::Network(format!(
            "Redfish unmount ISO HTTP {code}: {text}"
        )))
    }

    pub async fn set_boot(&self, target: BootTarget, once: bool) -> Result<(), AppError> {
        let boot_target = match target {
            BootTarget::Pxe => "Pxe",
            BootTarget::Disk => "Hdd",
        };
        let enabled = if once { "Once" } else { "Continuous" };
        let body = serde_json::json!({
            "BootSourceOverrideTarget": boot_target,
            "BootSourceOverrideEnabled": enabled,
            "BootSourceOverrideMode": "Legacy"
        });
        let resp = self
            .http
            .patch(self.system_url())
            .basic_auth(&self.username, Some(&self.password))
            .header(CONTENT_TYPE, "application/json")
            .header("OData-Version", "4.0")
            .json(&body)
            .send()
            .await
            .map_err(|e| AppError::Network(format!("Redfish set boot: {e}")))?;
        let code = resp.status().as_u16();
        if resp.status().is_success() || code == 204 {
            return Ok(());
        }
        let text = resp.text().await.unwrap_or_default();
        Err(AppError::Network(format!(
            "Redfish set boot HTTP {code}: {text}"
        )))
    }
}

File: frontend/src/lib/redfish.ts
// Browser Redfish BMC client. Mirrors backend/src/integration/bmc/redfish.rs so
// the JS oob-agent (running in the operator's browser) executes the same
// endpoints/actions as the on-network Rust client. Browsers cannot skip TLS
// validation, so self-signed BMCs surface as a TypeError the caller can detect.

export interface BmcCreds {
  address: string;
  username: string;
  password: string;
  redfishPath: string;
  tlsInsecure: boolean;
  preferred: string;
  timeoutSecs: number;
  ipmiInterface: string;
}

export type PowerState = 'on' | 'off' | 'unknown';

interface Ctx {
  base: string;
  systemPath: string;
  username: string;
  password: string;
  timeoutMs: number;
}

function basicAuth(username: string, password: string): string {
  return 'Basic ' + btoa(username + ':' + password);
}

function systemUrl(ctx: Ctx): string {
  const p = ctx.systemPath.replace(/\/+$/, '');
  return p.startsWith('http') ? p : ctx.base + p;
}

async function rfFetch(ctx: Ctx, url: string, init: RequestInit = {}): Promise<Response> {
  const controller = new AbortController();
  const t = setTimeout(() => controller.abort(), ctx.timeoutMs);
  try {
    return await fetch(url, {
      ...init,
      signal: controller.signal,
      headers: {
        Accept: 'application/json',
        'User-Agent': 'tcs-bmc/0.3',
        Authorization: basicAuth(ctx.username, ctx.password),
        ...(init.headers || {}),
      },
    });
  } finally {
    clearTimeout(t);
  }
}

async function readJson(res: Response): Promise<any> {
  const text = await res.text();
  try {
    return JSON.parse(text);
  } catch {
    return {};
  }
}

async function requireOk(res: Response, what: string): Promise<void> {
  if (!res.ok && res.status !== 204 && res.status !== 202) {
    const text = await res.text().catch(() => '');
    throw new Error(`Redfish ${what} HTTP ${res.status}: ${text}`.trim());
  }
}

export async function connect(creds: BmcCreds): Promise<Ctx> {
  const host = creds.address
    .replace(/^https?:\/\//, '')
    .replace(/\/+$/, '');
  const base = `https://${host}`;
  const timeoutMs = Math.max(5, creds.timeoutSecs) * 1000;

  const res = await rfFetch(ctx0(base, creds.username, creds.password, timeoutMs), `${base}/redfish/v1/Systems`);
  if (!res.ok) throw new Error(`Redfish Systems HTTP ${res.status}`);
  const body = await readJson(res);

  let systemPath: string;
  const rp = (creds.redfishPath || '').trim();
  if (rp) {
    systemPath = rp.startsWith('/') ? rp : `/${rp}`;
  } else {
    const first = body?.Members?.[0]?.['@odata.id'];
    if (!first) throw new Error('Redfish: no Systems members found');
    systemPath = first;
  }

  return { base, systemPath, username: creds.username, password: creds.password, timeoutMs };
}

function ctx0(base: string, username: string, password: string, timeoutMs: number): Ctx {
  return { base, systemPath: '', username, password, timeoutMs };
}

export async function getPowerState(ctx: Ctx): Promise<PowerState> {
  const res = await rfFetch(ctx, systemUrl(ctx));
  await requireOk(res, 'get system');
  const body = await readJson(res);
  const state = body?.PowerState;
  if (state === 'On') return 'on';
  if (state === 'Off') return 'off';
  return 'unknown';
}

export async function power(ctx: Ctx, action: string): Promise<void> {
  let resetType: string;
  switch (action) {
    case 'on':
      resetType = 'On';
      break;
    case 'off':
      resetType = 'ForceOff';
      break;
    case 'reset':
    case 'cycle':
      resetType = 'ForceRestart';
      break;
    case 'graceful_shutdown':
      resetType = 'GracefulShutdown';
      break;
    default:
      throw new Error(`Unknown power action: ${action}`);
  }
  const res = await rfFetch(ctx, `${systemUrl(ctx)}/Actions/ComputerSystem.Reset`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ ResetType: resetType }),
  });
  await requireOk(res, 'power');
}

export async function setBoot(ctx: Ctx, target: 'pxe' | 'disk', once: boolean): Promise<void> {
  const body = {
    BootSourceOverrideTarget: target === 'pxe' ? 'Pxe' : 'Hdd',
    BootSourceOverrideEnabled: once ? 'Once' : 'Continuous',
    BootSourceOverrideMode: 'Legacy',
  };
  const res = await rfFetch(ctx, systemUrl(ctx), {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json', 'OData-Version': '4.0' },
    body: JSON.stringify(body),
  });
  await requireOk(res, 'set boot');
}

async function findMedia(ctx: Ctx, media: string): Promise<{ odata: string; names: string[] }> {
  const res = await rfFetch(ctx, `${systemUrl(ctx)}/VirtualMedia`);
  await requireOk(res, 'get VirtualMedia');
  const body = await readJson(res);
  const members: any[] = body?.Members || [];
  if (members.length === 0) throw new Error('No VirtualMedia members found');
  const names = members.map((m) => m?.Name || '');
  const match = members.find((m) => (m?.Name || '').toLowerCase() === media.toLowerCase());
  if (!match) {
    throw new Error(`VirtualMedia '${media}' not found. Available: ${names.join(', ')}`);
  }
  return { odata: match['@odata.id'], names };
}

export async function mountIso(ctx: Ctx, isoUrl: string, media: string): Promise<void> {
  const { odata } = await findMedia(ctx, media);
  const detail = odata.startsWith('http') ? odata : ctx.base + odata;
  const res = await rfFetch(ctx, `${detail.replace(/\/+$/, '')}/VirtualMedia.InsertMedia`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      Image: isoUrl,
      TransferMethod: 'Download',
      TransferProtocolType: 'HTTP',
    }),
  });
  await requireOk(res, 'mount ISO');
}

export async function unmountIso(ctx: Ctx, media: string): Promise<void> {
  const { odata } = await findMedia(ctx, media);
  const detail = odata.startsWith('http') ? odata : ctx.base + odata;
  const res = await rfFetch(ctx, `${detail.replace(/\/+$/, '')}/VirtualMedia.EjectMedia`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
  });
  await requireOk(res, 'unmount ISO');
}
